x/pool-incentives/keeper/distr.go

```diff
@@ -0,0 +1,175 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/c-osmosis/osmosis/x/pool-incentives/types"
+)
+
+// GetAllocatableAsset gets the balance of the `MintedDenom` from the `feeCollectorName` module account and returns coins according to the `AllocationRatio`
+func (k Keeper) GetAllocatableAsset(ctx sdk.Context) sdk.Coin {
+	params := k.GetParams(ctx)
+
+	feeCollector := k.accountKeeper.GetModuleAccount(ctx, k.feeCollectorName)
+	asset := k.bankKeeper.GetBalance(ctx, feeCollector.GetAddress(), params.MintedDenom)
+
+	return sdk.NewCoin(asset.Denom, asset.Amount.ToDec().Mul(params.AllocationRatio).TruncateInt())
+}
+
+// AllocateAsset allocates and distributes coin according a pot’s proportional weight that is recorded in the record
+func (k Keeper) AllocateAsset(ctx sdk.Context, asset sdk.Coin) error {
+	distrInfo := k.GetDistrInfo(ctx)
+
+	if distrInfo.TotalWeight.GT(sdk.ZeroInt()) {
+		assetAmountDec := asset.Amount.ToDec()
+		totalWeightDec := distrInfo.TotalWeight.ToDec()
+		for _, record := range distrInfo.Records {
+			allocatingAmount := assetAmountDec.Mul(record.Weight.ToDec().Quo(totalWeightDec)).TruncateInt()
+			coins := sdk.NewCoins(sdk.NewCoin(asset.Denom, allocatingAmount))
+
+			err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, k.feeCollectorName, types.ModuleName, coins)
```

Comment:
If weight is very small, coins could be empty. And BeginBlocker calls panic if AllocateAsset returns error.